sycl/source/detail/graph_impl.cpp

```diff
@@ -61,76 +62,96 @@ void graph_impl::remove_root(const std::shared_ptr<node_impl> &Root) {
 // Recursive check if a graph node or its successors contains a given kernel
 // argument.
 //
-// @param[in] arg The kernel argument to check for.
-// @param[in] currentNode The current graph node being checked.
-// @param[in,out] deps The unique list of dependencies which have been
+// @param[in] Arg The kernel argument to check for.
+// @param[in] CurrentNode The current graph node being checked.
+// @param[in,out] Deps The unique list of dependencies which have been
 // identified for this arg.
-// @param[in] dereferencePtr if true arg comes direct from the handler in which
-// case it will need to be deferenced to check actual value.
 //
 // @returns True if a dependency was added in this node of any of its
 // successors.
 bool check_for_arg(const sycl::detail::ArgDesc &Arg,
                    const std::shared_ptr<node_impl> &CurrentNode,
-                   std::set<std::shared_ptr<node_impl>> &Deps,
-                   bool DereferencePtr = false) {
+                   std::set<std::shared_ptr<node_impl>> &Deps) {
   bool SuccessorAddedDep = false;
   for (auto &Successor : CurrentNode->MSuccessors) {
-    SuccessorAddedDep |= check_for_arg(Arg, Successor, Deps, DereferencePtr);
+    SuccessorAddedDep |= check_for_arg(Arg, Successor, Deps);
   }
 
-  if (Deps.find(CurrentNode) == Deps.end() &&
-      CurrentNode->has_arg(Arg, DereferencePtr) && !SuccessorAddedDep) {
+  if (Deps.find(CurrentNode) == Deps.end() && CurrentNode->has_arg(Arg) &&
+      !SuccessorAddedDep) {
     Deps.insert(CurrentNode);
     return true;
   }
   return SuccessorAddedDep;
 }
 
-template <typename T>
 std::shared_ptr<node_impl>
-graph_impl::add(const std::shared_ptr<graph_impl> &impl, T cgf,
-                const std::vector<sycl::detail::ArgDesc> &args,
-                const std::vector<std::shared_ptr<node_impl>> &dep) {
-  std::shared_ptr<node_impl> nodeImpl =
-      std::make_shared<node_impl>(impl, cgf, args);
+graph_impl::add(const std::shared_ptr<graph_impl> &Impl,
+                std::function<void(handler &)> CGF,
+                const std::vector<sycl::detail::ArgDesc> &Args,
+                const std::vector<std::shared_ptr<node_impl>> &Dep) {
+  sycl::queue TempQueue{};
```

Comment:
Is this logic going to be reworked once the device is known at this point? If so we might want to add a note with a TODO. Looks like, as is, there's a risk of ignoring properties of a queue that will be assigned in future, or are we considering to update them at a later point?

Reply:
If we change the API to set a device, we should create `tempQueue` with that device. However, there is no way of knowing during graph creation time the properties of a queue an executable graph will be submitted to,
as we only finalize on a context.

The two queue properties defined by the spec are 1. in-order 2. profiling
If a graph is going to be queue property aware during construction then I think it should be the complete graph, I don't think it makes sense to have a single node be in-order or profileable, and not the rest of the graph.

So we could define properties to the command_graph constructor that would specify the queue properties that get set here in tempQueue. However, there's still no guarantee that the queue submitted to will match these, unless we specify that it's an error to submit an executable graph to a queue with different properties than the graph was created with.

Reply:
I looked into providing a new `sycl::handler` constructor rather than creating a temporary queue. It seems to work okay, but created a separate PR for that https://github.com/reble/llvm/pull/101 targetting this branch since I'm not sure about it and would like to chat with Ben about it before pushing.

Reply:
Okay sounds good to me!